File: PycharmProjects/contract_utilities/contract_meta_info.py
__author__ = 'kocat_000'

import math as m
import pytz as pytz
import datetime as dt

full_letter_month_list = ['F', 'G', 'H', 'J', 'K', 'M', 'N', 'Q', 'U', 'V', 'X', 'Z']
quarterly_month_list = ['H', 'M', 'U', 'Z']
letter_month_string = 'FGHJKMNQUVXZ'
central_zone = pytz.timezone('US/Central')

futures_contract_months = {'LN': ['G', 'J', 'K', 'M', 'N', 'Q', 'V', 'Z'],
                           'LC': ['G', 'J', 'M', 'Q', 'V', 'Z'],
                           'FC': ['F', 'H', 'J', 'K', 'Q', 'U', 'V', 'X'],
                           'C': ['H', 'K', 'N', 'U', 'Z'],
                           'S': ['F', 'H', 'K', 'N', 'Q', 'U', 'X'],
                           'SM': ['F', 'H', 'K', 'N', 'Q', 'U', 'V', 'Z'],
                           'BO': ['F', 'H', 'K', 'N', 'Q', 'U', 'V', 'Z'],
                           'W': ['H', 'K', 'N', 'U', 'Z'],
                           'KW': ['H', 'K', 'N', 'U', 'Z'],
                           'SB': ['H', 'K', 'N', 'V'],
                           'KC': ['H', 'K', 'N', 'U', 'Z'],
                           'CC': ['H', 'K', 'N', 'U', 'Z'],
                           'CT': ['H', 'K', 'N', 'V', 'Z'],
                           'OJ': ['F', 'H', 'K', 'N', 'U', 'X'],
                           'ED': full_letter_month_list,
                           'CL': full_letter_month_list,
                           'B' : full_letter_month_list,
                           'HO': full_letter_month_list,
                           'RB': full_letter_month_list,
                           'NG': full_letter_month_list,
                           'ES': quarterly_month_list,
                           'NQ': quarterly_month_list,
                           'EC': quarterly_month_list,
                           'JY': quarterly_month_list,
                           'AD': quarterly_month_list,
                           'CD': quarterly_month_list,
                           'BP': quarterly_month_list,
                           'TY': quarterly_month_list,
                           'US': quarterly_month_list,
                           'FV': quarterly_month_list,
                           'TU': quarterly_month_list,
                           'GC': full_letter_month_list,
                           'SI': full_letter_month_list}

option_cabinet_values = {'ED': 0.0025,
                         'E0': 0.0025,
                         'E2': 0.0025,
                         'E3': 0.0025,
                         'E4': 0.0025,
                         'E5': 0.0025,
                         'LC': 0.0125,
                         'LN': 0.0125,
                         'FC': 0.0125,
                         'ES': 0.05,
                         'NQ': 0.05,
                         'TU': 1/128,
                         'FV': 1/128,
                         'TY': 1/64,
                         'US': 1/64,
                         'EC': 0.00005,
                         'JY': 5,
                         'AD': 0.00005,
                         'CD': 0.00005,
                         'BP': 0.0001}


def get_option_exercise_type(**kwargs):

    ticker_head = kwargs['ticker_head']

    if ticker_head in ['LN', 'LC', 'ES', 'EC', 'JY', 'AD', 'CD', 'BP', 'GC', 'SI',
                       'TY', 'US', 'FV', 'TU', 'C', 'S', 'SM', 'BO', 'W', 'CL', 'NG',
                       'ED', 'E0', 'E2', 'E3', 'E4', 'E5']:
        exercise_type = 'A'
    else:
        exercise_type = None

    return exercise_type


def get_option_contract_months(**kwargs):

    ticker_head = kwargs['ticker_head']
    contract_months = []

    if ticker_head == 'LN':
        contract_months = ['G', 'J', 'K', 'M', 'N', 'Q', 'V', 'Z']
    elif ticker_head in option_tickerhead_list:
        contract_months = full_letter_month_list

    return contract_months

futures_butterfly_strategy_tickerhead_list = ['LN', 'LC', 'FC',
                                              'C', 'S', 'SM', 'BO', 'W', 'KW',
                                              'SB', 'KC', 'CC', 'CT', 'OJ',
                                              'CL', 'B', 'HO', 'RB', 'NG', 'ED']

cme_futures_tickerhead_list = ['LN', 'LC', 'FC',
                               'C', 'S', 'SM', 'BO', 'W', 'KW',
                               'CL', 'HO', 'RB', 'NG', 'ED',
                               'ES', 'NQ', 'EC', 'JY', 'AD', 'CD', 'BP',
                               'TY', 'US', 'FV', 'TU', 'GC', 'SI']

cme_option_tickerhead_list = ['LN', 'LC', 'ES', 'EC', 'JY', 'AD', 'CD', 'BP', 'GC', 'SI',
                              'TY', 'US', 'FV', 'TU', 'C', 'S', 'SM', 'BO', 'W', 'CL', 'NG',
                              'ED', 'E0', 'E2', 'E3', 'E4', 'E5']

option_tickerhead_list = cme_option_tickerhead_list


def get_tickerhead_list(list_name):

    if list_name == 'butterfly':
        tickerhead_list = futures_butterfly_strategy_tickerhead_list
    elif list_name == 'cme_futures':
        tickerhead_list = cme_futures_tickerhead_list
    elif list_name == 'cme_options':
        tickerhead_list = cme_option_tickerhead_list

    return tickerhead_list


def get_exchange_traded(ticker_head):

    if (ticker_head in cme_futures_tickerhead_list) or (ticker_head in cme_option_tickerhead_list):
        exchange = 'CME'
    elif ticker_head in futures_butterfly_strategy_tickerhead_list:
        exchange = 'ICE'
    else:
        exchange = null

    return exchange

contract_name = {'LN': 'Lean Hog',
                 'LC': 'Live Cattle',
                 'FC': 'Feeder Cattle',
                 'C': 'Corn',
                 'S': 'Soybean',
                 'SM': 'Soybean Meal',
                 'BO': 'Soybean Oil',
                 'W': 'Chicago SRW Wheat',
                 'KW': 'Kansas HRW Wheat',
                 'SB': 'Sugar 11',
                 'KC': 'Coffee C',
                 'CC': 'Cocoa',
                 'CT': 'Cotton No. 2',
                 'OJ': 'FCOJ-A',
                 'CL': 'Crude Oil',
                 'B' : 'Brent Crude',
                 'HO': 'NY Harbor ULSD',
                 'RB': 'RBOB Gasoline',
                 'NG': 'Henry Hub Natural Gas',
                 'ED': 'Eurodollar',
                 'E0': 'Eurodollar 1 Year Mid-Curve Options',
                 'E2': 'Eurodollar 2 Year Mid-Curve Options',
                 'E3': 'Eurodollar 3 Year Mid-Curve Options',
                 'E4': 'Eurodollar 4 Year Mid-Curve Options',
                 'E5': 'Eurodollar 5 Year Mid-Curve Options',
                 'ES': 'E-mini S&P 500',
                 'NQ': 'E-mini Nasdaq 100',
                 'EC': 'Euro FX',
                 'JY': 'Japanese Yen',
                 'AD': 'Australian Dollar',
                 'CD': 'Canadian Dollar',
                 'BP': 'British Pound',
                 'TY': '10-Year T-Note',
                 'US': 'US Treasury Bond',
                 'FV': '5-Year T-Note',
                 'TU': '2-Year T-Note',
                 'GC': 'Gold',
                 'SI': 'Silver'}

relevant_max_cal_dte = {'LN': 720,
                        'LC': 720,
                        'FC': 720,
                        'C': 720,
                        'S': 720,
                        'SM': 720,
                        'BO': 720,
                        'W': 720,
                        'KW': 720,
                        'SB': 720,
                        'KC': 720,
                        'CC': 720,
                        'CT': 720,
                        'OJ': 720,
                        'CL': 720,
                        'B' : 720,
                        'HO': 720,
                        'RB': 720,
                        'NG': 720,
                        'ED': 1440}

last_trade_hour_minute = {'LN': dt.time(13, 3, 0, 0),
                          'LC': dt.time(13, 3, 0, 0),
                          'FC': dt.time(13, 3, 0, 0),
                          'C': dt.time(13, 18, 0, 0),
                          'S': dt.time(13, 18, 0, 0),
                          'SM': dt.time(13, 18, 0, 0),
                          'BO': dt.time(13, 18, 0, 0),
                          'W': dt.time(13, 18, 0, 0),
                          'KW': dt.time(13, 18, 0, 0),
                          'SB': dt.time(11, 58, 0, 0),
                          'KC': dt.time(12, 28, 0, 0),
                          'CC': dt.time(12, 28, 0, 0),
                          'CT': dt.time(12, 28, 0, 0),
                          'OJ': dt.time(12, 58, 0, 0),
                          'CL': dt.time(14, 28, 0, 0),
                          'B': dt.time(14, 28, 0, 0),
                          'HO': dt.time(14, 28, 0, 0),
                          'RB': dt.time(14, 28, 0, 0),
                          'NG': dt.time(14, 28, 0, 0),
                          'ED': dt.time(14, 28, 0, 0),
                          'ES': dt.time(14, 28, 0, 0),
                          'NQ': dt.time(14, 28, 0, 0),
                          'EC': dt.time(14, 28, 0, 0),
                          'JY': dt.time(14, 28, 0, 0),
                          'AD': dt.time(14, 28, 0, 0),
                          'CD': dt.time(14, 28, 0, 0),
                          'BP': dt.time(14, 28, 0, 0),
                          'TY': dt.time(14, 28, 0, 0),
                          'US': dt.time(14, 28, 0, 0),
                          'FV': dt.time(14, 28, 0, 0),
                          'TU': dt.time(14, 28, 0, 0),
                          'GC': dt.time(14, 28, 0, 0),
                          'SI': dt.time(14, 28, 0, 0)}


first_trade_hour_minute = {'LN': dt.time(8, 31, 0, 0),
                           'LC': dt.time(8, 31, 0, 0),
                           'FC': dt.time(8, 31, 0, 0),
                           'C': dt.time(8, 31, 0, 0),
                           'S': dt.time(8, 31, 0, 0),
                           'SM': dt.time(8, 31, 0, 0),
                           'BO': dt.time(8, 31, 0, 0),
                           'W': dt.time(8, 31, 0, 0),
                           'KW': dt.time(8, 31, 0, 0),
                           'SB': dt.time(3, 0, 0, 0),
                           'KC': dt.time(4, 0, 0, 0),
                           'CC': dt.time(4, 0, 0, 0),
                           'CT': dt.time(0, 45, 0, 0),
                           'OJ': dt.time(8, 0, 0, 0),
                           'CL': dt.time(0, 45, 0, 0),
                           'B': dt.time(0, 45, 0, 0),
                           'HO': dt.time(0, 45, 0, 0),
                           'RB': dt.time(0, 45, 0, 0),
                           'NG': dt.time(0, 45, 0, 0),
                           'ED': dt.time(0, 45, 0, 0),
                           'ES': dt.time(0, 45, 0, 0),
                           'NQ': dt.time(0, 45, 0, 0),
                           'EC': dt.time(0, 45, 0, 0),
                           'JY': dt.time(0, 45, 0, 0),
                           'AD': dt.time(0, 45, 0, 0),
                           'CD': dt.time(0, 45, 0, 0),
                           'BP': dt.time(0, 45, 0, 0),
                           'TY': dt.time(0, 45, 0, 0),
                           'US': dt.time(0, 45, 0, 0),
                           'FV': dt.time(0, 45, 0, 0),
                           'TU': dt.time(0, 45, 0, 0),
                           'GC': dt.time(0, 45, 0, 0),
                           'SI': dt.time(0, 45, 0, 0)}


def get_max_cal_dte(**kwargs):

    ticker_head = kwargs['ticker_head']
    ticker_month = kwargs['ticker_month']

    if ticker_head in ['LN', 'LC', 'FC', 'C', 'S', 'SM', 'BO', 'W', 'KW', 'SB', 'KC', 'CC', 'CT', 'OJ', 'HO', 'RB', 'NG']:
        max_cal_dte = 720
    elif ticker_head in ['ED', 'E0','E2', 'E3', 'E4', 'E5']:
        max_cal_dte = 1440
    elif ticker_head == 'CL':
        if ticker_month in [6, 12]:
            max_cal_dte = 1080
        else:
            max_cal_dte = 720
    elif ticker_head == 'B':
        if ticker_month in [6, 12]:
            max_cal_dte = 1440
        else:
            max_cal_dte = 720
    elif ticker_head in ['ES', 'NQ', 'TY', 'US', 'FV', 'TU', 'GC', 'SI']:
        max_cal_dte = 360
    elif ticker_head in ['EC', 'JY', 'AD', 'CD', 'BP']:
        max_cal_dte = 720

    return max_cal_dte

ticker_class = {'LN': 'Livestock',
                'LC': 'Livestock',
                'FC': 'Livestock',
                'ES': 'Index',
                'NQ': 'Index',
                'EC': 'FX',
                'JY': 'FX',
                'AD': 'FX',
                'CD': 'FX',
                'BP': 'FX',
                'GC': 'Metal',
                'SI': 'Metal',
                'TY': 'Treasury',
                'US': 'Treasury',
                'FV': 'Treasury',
                'TU': 'Treasury',
                'C': 'Ag',
                'S': 'Ag',
                'SM': 'Ag',
                'BO': 'Ag',
                'W': 'Ag',
                'KW': 'Ag',
                'SB': 'Soft',
                'KC': 'Soft',
                'CC': 'Soft',
                'CT': 'Soft',
                'OJ': 'Soft',
                'CL': 'Energy',
                'B' : 'Energy',
                'HO': 'Energy',
                'RB': 'Energy',
                'NG': 'Energy',
                'ED': 'STIR',
                'E0': 'STIR',
                'E2': 'STIR',
                'E3': 'STIR',
                'E4': 'STIR',
                'E5': 'STIR'}

aligned_data_tickerhead = {'LN': 'LNH',
                           'LC': 'LVC',
                           'ES': 'SP',
                           'EC': 'EC',
                           'JY': 'JY100',
                           'AD': 'AD',
                           'CD': 'C1',
                           'BP': 'BP1',
                           'GC': 'GC',
                           'SI': 'SI',
                           'TY': 'TY',
                           'US': 'US',
                           'FV': 'FV',
                           'TU': 'TU',
                           'C': 'CB',
                           'S': 'SB',
                           'SM': 'SM',
                           'BO': 'BO',
                           'W': 'WW',
                           'CL': 'CL',
                           'NG': 'NG',
                           'ED': 'ED',
                           'E0': 'E1',
                           'E2': 'E2',
                           'E3': 'E3',
                           'E4': 'E4',
                           'E5': 'E5'}

aligned_data_tr_dte_list = [10, 15, 20, 25, 30, 35, 40, 45, 50, 60, 70, 80, 90, 100,
                           120, 140, 160, 180, 200, 220, 240, 260, 280, 300, 330, 360, 390]

contract_multiplier = {'LN': 400,
                       'LC': 400,
                       'FC': 500,
                       'C': 50,
                       'S': 50,
                       'SM': 100,
                       'BO': 600,
                       'W': 50,
                       'KW': 50,
                       'SB': 1120,
                       'KC': 375,
                       'CC': 10,
                       'CT': 500,
                       'OJ': 150,
                       'CL': 1000,
                       'B': 1000,
                       'HO': 42000,
                       'RB': 42000,
                       'NG': 10000,
                       'ED': 2500,
                       'E0': 2500,
                       'E2': 2500,
                       'E3': 2500,
                       'E4': 2500,
                       'E5': 2500,
                       'ES': 50,
                       'NQ': 20,
                       'AD': 100000,
                       'CD': 100000,
                       'EC': 125000,
                       'JY': 1.25,
                       'BP': 62500,
                       'FV': 1000,
                       'TU': 2000,
                       'TY': 1000,
                       'US': 1000,
                       'GC': 100,
                       'SI': 5000}
t_cost = {'CL': 0.80,
          'NG': 0.78,
          'B': 0.93,
          'HO': 0.80,
          'LC': 0.61,
          'LN': 0.61,
          'FC': 0.61,
          'RB': 0.80,
          'C': 0.61,
          'W': 0.61,
          'S': 0.61,
          'SM': 0.61,
          'BO': 0.61,
          'KW': 0.61,
          'ED': 0.29,
          'JY': 0.42,
          'EC': 0.42,
          'BP': 0.42,
          'AD': 0.42,
          'CD': 0.42,
          'TY': 0.31,
          'FV': 0.31,
          'TU': 0.31,
          'ES': 0.45,
          'NQ': 0.45,
          'SI': 0.65,
          'GC': 0.65,
          'KC': 2.1,
          'CC': 2.1,
          'SB': 2.1,
          'OJ': 2.1,
          'CT': 2.1}


def get_contract_specs(ticker):
    return {'ticker_head': ticker[:-5],
            'ticker_year': int(ticker[-4:]),
            'ticker_month_str': ticker[-5],
            'ticker_month_num': letter_month_string.find(ticker[-5])+1,
            'ticker_class': ticker_class[ticker[:-5]],
            'cont_indx': 100*int(ticker[-4:]) + letter_month_string.find(ticker[-5])+1}


def get_month_seperation_from_cont_indx(cont_indx1, cont_indx2):

    contract_month1 = cont_indx1%100
    contract_month2 = cont_indx2%100
    contract_year1 = m.floor(cont_indx1/100)
    contract_year2 = m.floor(cont_indx2/100)

    return 12*(contract_year1-contract_year2)+(contract_month1-contract_month2)


def get_cont_indx_from_month_seperation(cont_indx1,month_seperation):

    contract_month1 = cont_indx1 % 100
    contract_year1 = round(cont_indx1/100)

    total_months1 = 12*contract_year1 + contract_month1

    total_months2 = total_months1+month_seperation

    contract_month2 = total_months2 % 12

    if contract_month2 == 0:
        contract_month2 = 12
        contract_year2 = round(total_months2/12)-1
    else:
        contract_year2 = m.floor(total_months2/12)

    return 100*contract_year2+contract_month2

File: c#/ContractUtilities/ContractUtilities/ContractMetaInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ContractUtilities
{
    public class ContractSpecs
    {
        public string tickerHead { get; set;}
        public int tickerYear { get; set; }
        public string tickerMonthStr { get; set; }
        public int tickerMonthNum { get; set; }
        public string tickerClass { get; set; }
        public int contINDX { get; set; }
    }

    public static class ContractMetaInfo
    {
        public static List<string> FuturesButterflyTickerheadList = 
            new List<string> (new string[] {"LN", "LC", "FC",
                                              "C", "S", "SM", "BO", "W", "KW",
                                              "SB", "KC", "CC", "CT", "OJ",
                                              "CL", "B", "HO", "RB", "NG", "ED"});

        public static List<string> cmeFuturesTickerheadList = 
            new List<string> (new string[] {"LN", "LC", "FC",
                               "C", "S", "SM", "BO", "W", "KW",
                               "CL", "HO", "RB", "NG", "ED",
                               "ES", "NQ", "EC", "JY", "AD", "CD", "BP",
                               "TY", "US", "FV", "TU", "GC", "SI"});

        public static List<string> AmericanExerciseTickerheadList =
            new List<string>(new string[] {"LN", "LC", "ES", "EC", "JY", "AD", "CD", "BP", "GC", "SI",
                       "TY", "US", "FV", "TU", "C", "S", "SM", "BO", "W", "CL", "NG",
                       "ED", "E0", "E2", "E3", "E4", "E5" });

        public static List<string> iceFuturesTickerheadList =
            new List<string>(new string[] { "SB", "KC", "CC", "CT", "OJ", "B" });

        public static Dictionary<string, string> tickerClassDict = new Dictionary<string, string>
            {
            {"LN","Livestock"},{"LC","Livestock"},{"FC","Livestock"},{"ES","Index"},{"NQ","Index"},
            {"EC","FX"},{"JY","FX"},{"AD","FX"},{"CD","FX"},{"BP","FX"},{"GC","Metal"},{"SI","Metal"},
            {"TY","Treasury"},{"US","Treasury"},{"FV","Treasury"},{"TU","Treasury"},
            {"C","Ag"},{"S","Ag"},{"SM","Ag"},{"BO","Ag"},{"W","Ag"},{"KW","Ag"},
            {"SB","Soft"},{"KC","Soft"},{"CC","Soft"},{"CT","Soft"},{"OJ","Soft"},
            {"CL","Energy"},{"B","Energy"},{"HO","Energy"},{"RB","Energy"},{"NG","Energy"},
            {"ED","STIR"},{"E0","STIR"},{"E2","STIR"},{"E3","STIR"},{"E4","STIR"},{"E5","STIR"}
            };

        public static List<string> fullLetterMonthList = new List<string> { "F", "G", "H", "J", "K", "M", "N", "Q", "U", "V", "X", "Z" };
        public static List<string> QuarterlyLetterMonthList = new List<string> { "H", "M", "U", "Z" };

        public static string GetExchange4Tickerhead(string tickerHead)
        {
            if (cmeFuturesTickerheadList.Contains(tickerHead))
                return "CME";
            else if (iceFuturesTickerheadList.Contains(tickerHead))
                return "ICE";
            else
                return null;
        }

        public static Dictionary<string, List<string>> FuturesContractMonths = new Dictionary<string, List<string>>
        {{"LN",new List<string>(new string[] {"G", "J", "K", "M", "N", "Q", "V", "Z"})},
         {"LC",new List<string>(new string[] {"G", "J", "M", "Q", "V", "Z"})},
         {"FC",new List<string>(new string[] {"F", "H", "J", "K", "Q", "U", "V", "X"})},
         {"C",new List<string>(new string[] {"H", "K", "N", "U", "Z"})},
         {"S",new List<string>(new string[] {"F", "H", "K", "N", "Q", "U", "X"})},
         {"SM",new List<string>(new string[] {"F", "H", "K", "N", "Q", "U", "V", "Z"})},
         {"BO",new List<string>(new string[] {"F", "H", "K", "N", "Q", "U", "V", "Z"})},
         {"W",new List<string>(new string[] {"H", "K", "N", "U", "Z"})},
         {"KW",new List<string>(new string[] {"H", "K", "N", "U", "Z"})},
         {"SB",new List<string>(new string[] {"H", "K", "N", "V"})},
         {"KC",new List<string>(new string[] {"H", "K", "N", "U", "Z"})},
         {"CC",new List<string>(new string[] {"H", "K", "N", "U", "Z"})},
         {"CT",new List<string>(new string[] {"H", "K", "N", "V", "Z"})},
         {"OJ",new List<string>(new string[] {"F", "H", "K", "N", "U", "X"})},
         {"ED",fullLetterMonthList},
         {"CL",fullLetterMonthList},
         {"B",fullLetterMonthList},
         {"HO",fullLetterMonthList},
         {"RB",fullLetterMonthList},
         {"NG",fullLetterMonthList},
         {"ES",QuarterlyLetterMonthList},
         {"NQ",QuarterlyLetterMonthList},
         {"EC",QuarterlyLetterMonthList},
         {"JY",QuarterlyLetterMonthList},
         {"AD",QuarterlyLetterMonthList},
         {"CD",QuarterlyLetterMonthList},
         {"BP",QuarterlyLetterMonthList},
         {"TY",QuarterlyLetterMonthList},
         {"US",QuarterlyLetterMonthList},
         {"FV",QuarterlyLetterMonthList},
         {"TU",QuarterlyLetterMonthList},
         {"GC",fullLetterMonthList},
         {"SI",fullLetterMonthList}};

        public static Dictionary<string, double> ContractMultiplier = new Dictionary<string, double>
        {{"LN", 400},
         {"LC", 400},
         {"FC", 500},
         {"C", 50},
         {"S", 50},
         {"SM", 100},
         {"BO", 600},
         {"W", 50},
         {"KW", 50},
         {"SB", 1120},
         {"KC", 375},
         {"CC", 10},
         {"CT", 500},
         {"OJ", 150},
         {"CL", 1000},
         {"B", 1000},
         {"HO", 42000},
         {"RB", 42000},
         {"NG", 10000},
         {"ED", 2500},
         {"E0", 2500},
         {"E2", 2500},
         {"E3", 2500},
         {"E4", 2500},
         {"E5", 2500},
         {"ES", 50},
         {"NQ", 20},
         {"AD", 100000},
         {"CD", 100000},
         {"EC", 125000},
         {"JY", 1.25},
         {"BP", 62500},
         {"FV", 1000},
         {"TU", 2000},
         {"TY", 1000},
         {"US", 1000},
         {"GC", 100},
         {"SI", 5000}};

        public static string GetOptionExerciseType(string tickerHead)
        {
            if (AmericanExerciseTickerheadList.Contains(tickerHead))
            {
                return "A";
            }
            else
            {
                return "";
            }
        }


        public static ContractSpecs GetContractSpecs(string ticker)
        {
            ContractSpecs specsOutput = new ContractSpecs();
            specsOutput.tickerHead = ticker.Substring(0, ticker.Count() - 5);
            specsOutput.tickerYear = Convert.ToInt16(ticker.Substring(ticker.Count() - 4, 4));
            specsOutput.tickerMonthStr = ticker.Substring(ticker.Count() - 5, 1);
            specsOutput.tickerMonthNum = fullLetterMonthList.FindIndex(x => x == specsOutput.tickerMonthStr) + 1;
            specsOutput.tickerClass = tickerClassDict[specsOutput.tickerHead];
            specsOutput.contINDX = 100*specsOutput.tickerYear + specsOutput.tickerMonthNum;
            return specsOutput;
        }      
    }
}
